framework/src/systems/NonlinearSystemBase.C

```diff
@@ -1553,25 +1553,31 @@ NonlinearSystemBase::computeResidualInternal(const std::set<TagID> & tags)
   // residual contributions from the domain
   PARALLEL_TRY
   {
-    TIME_SECTION("Kernels", 3 /*, "Computing Kernels"*/);
+    if (_kernels.hasActiveObjects() || _dg_kernels.hasActiveObjects() ||
+        _interface_kernels.hasActiveObjects() || _integrated_bcs.hasActiveObjects() ||
+        _fe_problem.haveFV())
+    {
+      TIME_SECTION("Kernels", 3 /*, "Computing Kernels"*/);
 
-    ConstElemRange & elem_range = *_mesh.getActiveLocalElementRange();
+      ConstElemRange & elem_range = *_mesh.getActiveLocalElementRange();
 
-    ComputeResidualThread cr(_fe_problem, tags);
-    Threads::parallel_reduce(elem_range, cr);
+      ComputeResidualThread cr(_fe_problem, tags);
```

Comment:
Maybe move that check inside `NonlinearThread::hasObjects`? I feel it is safer in this way because it sees directly all the objects. We can still build this `ComputeResidualThread`, but will call parallel_reduce conditionally.

Reply:
Yeah, at the minimum. I'm still a bit uncomfortable about this. What's so bad about looping through the elements to realize we have no objects?

Reply:
I don't understand why we need to loop through the mesh... This is at least obvious about what we are skipping. We have similar logic below for nodal kernels.

Reply:
> What's so bad about looping through the elements to realize we have no objects?

`O(N_elem)` is asymptotically worse than `O(0)`?  Even on DistributedMesh the element loop is something like `O(N_elem/N_proc * log(N_elem/N_proc))`; we don't have to predicate our way past a zillion non-local elements but we do have to iterate through an ordered map (or a more complicated data structure that has a better constant but the same asymptotics) under the hood.

Reply:
> I don't understand why we need to loop through the mesh... This is at least obvious about what we are skipping. We have similar logic below for nodal kernels.

Because it is not immediately clear if we add capability to the threaded loop object to execute more objects, but then forget to update this change.

I'm not completely against this -  but this logic should at the very least be local to the threaded loop object, not a System method.

Reply:
I like the idea of adding an early exit to the threaded object if there is nothing to do.

Reply:
So, the approach what you guys agree is to implement a `hasActiveObjects` function inside `NonlinearThread` and use that to skip the calculation?

Reply:
Or add a check in `onElement` on the existence of objects like other routines like `onInternalSide` and do an empty loop?

Reply:
Once MOOSE maintainers confirm, I will push it again.